### PR TITLE
ETI Implementation for DeepCopy and ViewFill: Issue 1578

### PR DESCRIPTION
**Pull request by crtrott**

### CODE REVIEW

File: core/src/Kokkos_CopyViews.hpp [hunk 1]
```diff
@@ -1298,7 +1490,9 @@ void deep_copy
     Kokkos::fence();
   } else {
     Kokkos::fence();
-    Impl::view_copy(dst,src);
+  printf("ViewCopyEnter\n");
```
Comment (ibaned) [inline core/src/Kokkos_CopyViews.hpp]: Take those out

**Comment by ibaned** [comment]
There are also compile failures in Travis:
```
/home/travis/build/kokkos/kokkos/core/src/Kokkos_CopyViews.hpp: In instantiation of ‘void Kokkos::deep_copy(const ExecSpace&, const Kokkos::View<D, P ...>&, const Kokkos::View<ST, SP ...>&, typename std::enable_if<(((Kokkos::is_execution_space<ExecutionSpace>::value && std::is_same<typename Kokkos::ViewTraits<D, P ...>::specialize, void>::value) && std::is_same<typename Kokkos::ViewTraits<ST, SP ...>::specialize, void>::value) && (((unsigned int)(Kokkos::ViewTraits<D, P ...>::rank) != 0) || ((unsigned int)(Kokkos::ViewTraits<ST, SP ...>::rank) != 0)))>::type*) [with ExecSpace = Kokkos::Serial; DT = double* [3][5][7]; DP = {Kokkos::Serial}; ST = double* [3][5][7]; SP = {Kokkos::LayoutRight, Kokkos::Serial}; typename std::enable_if<(((Kokkos::is_execution_space<ExecutionSpace>::value && std::is_same<typename Kokkos::ViewTraits<D, P ...>::specialize, void>::value) && std::is_same<typename Kokkos::ViewTraits<ST, SP ...>::specialize, void>::value) && (((unsigned int)(Kokkos::ViewTraits<D, P ...>::rank) != 0) || ((unsigned int)(Kokkos::ViewTraits<ST, SP ...>::rank) != 0)))>::type = void]’:
/home/travis/build/kokkos/kokkos/core/unit_test/TestViewAPI.hpp:1074:69:   required from ‘static void Test::TestViewAPI<T, DeviceType>::run_test() [with T = double; DeviceType = Kokkos::Serial]’
/home/travis/build/kokkos/kokkos/core/unit_test/TestViewAPI_a.hpp:51:42:   required from here
/home/travis/build/kokkos/kokkos/core/src/Kokkos_CopyViews.hpp:1741:66: error: dependent-name ‘dst_type:: uniform_runtime_nomemspace_type’ is parsed as a non-type, but instantiation yields a type
     Impl::view_copy(dst_type::uniform_runtime_nomemspace_type(dst),src_type::uniform_runtime_const_nomemspace_type(src));
```

**Comment by ibaned** [comment]
Left out the last line of the compile error that shows how to fix it:
```
In file included from /home/travis/build/kokkos/kokkos/core/src/Kokkos_Core.hpp:93:0,
                 from /home/travis/build/kokkos/kokkos/core/unit_test/TestViewAPI.hpp:46,
                 from /home/travis/build/kokkos/kokkos/core/unit_test/TestViewAPI_a.hpp:44,
                 from /home/travis/build/kokkos/kokkos/core/unit_test/serial/TestSerial_ViewAPI_a.cpp:45:
/home/travis/build/kokkos/kokkos/core/src/Kokkos_CopyViews.hpp: In instantiation of ‘void Kokkos::deep_copy(const ExecSpace&, const Kokkos::View<D, P ...>&, const Kokkos::View<ST, SP ...>&, typename std::enable_if<(((Kokkos::is_execution_space<ExecutionSpace>::value && std::is_same<typename Kokkos::ViewTraits<D, P ...>::specialize, void>::value) && std::is_same<typename Kokkos::ViewTraits<ST, SP ...>::specialize, void>::value) && (((unsigned int)(Kokkos::ViewTraits<D, P ...>::rank) != 0) || ((unsigned int)(Kokkos::ViewTraits<ST, SP ...>::rank) != 0)))>::type*) [with ExecSpace = Kokkos::Serial; DT = double* [3][5][7]; DP = {Kokkos::Serial}; ST = double* [3][5][7]; SP = {Kokkos::LayoutRight, Kokkos::Serial}; typename std::enable_if<(((Kokkos::is_execution_space<ExecutionSpace>::value && std::is_same<typename Kokkos::ViewTraits<D, P ...>::specialize, void>::value) && std::is_same<typename Kokkos::ViewTraits<ST, SP ...>::specialize, void>::value) && (((unsigned int)(Kokkos::ViewTraits<D, P ...>::rank) != 0) || ((unsigned int)(Kokkos::ViewTraits<ST, SP ...>::rank) != 0)))>::type = void]’:
/home/travis/build/kokkos/kokkos/core/unit_test/TestViewAPI.hpp:1074:69:   required from ‘static void Test::TestViewAPI<T, DeviceType>::run_test() [with T = double; DeviceType = Kokkos::Serial]’
/home/travis/build/kokkos/kokkos/core/unit_test/TestViewAPI_a.hpp:51:42:   required from here
/home/travis/build/kokkos/kokkos/core/src/Kokkos_CopyViews.hpp:1741:66: error: dependent-name ‘dst_type:: uniform_runtime_nomemspace_type’ is parsed as a non-type, but instantiation yields a type
     Impl::view_copy(dst_type::uniform_runtime_nomemspace_type(dst),src_type::uniform_runtime_const_nomemspace_type(src));
                                                                  ^
/home/travis/build/kokkos/kokkos/core/src/Kokkos_CopyViews.hpp:1741:66: note: say ‘typename dst_type:: uniform_runtime_nomemspace_type’ if a type is meant
```
Just missing `typename`.

**Comment by crtrott** [comment]
```
Running on machine: apollo
WARNING!! THE FOLLOWING CHANGES ARE UNCOMMITTED!! :
?? core/src/diff_cuda_large
?? remove_deep_copy_impl

Repository Status:  d1b2be6d2d76757f8c6dafc1e1a9a904f8ff5e14 Remove printf


Going to test compilers:  gcc/4.8.4 gcc/5.3.0 intel/16.0.1 clang/3.9.0 clang/6.0 cuda/9.1
Testing compiler gcc/4.8.4
  Starting job gcc-4.8.4-OpenMP-release
  PASSED gcc-4.8.4-OpenMP-release
Testing compiler gcc/5.3.0
  Starting job gcc-4.8.4-Pthread-release
  PASSED gcc-4.8.4-Pthread-release
Testing compiler intel/16.0.1
  Starting job gcc-5.3.0-Serial-release
  PASSED gcc-5.3.0-Serial-release
Testing compiler clang/3.9.0
  Starting job intel-16.0.1-OpenMP-release
  PASSED intel-16.0.1-OpenMP-release
Testing compiler clang/6.0
  Starting job clang-3.9.0-Pthread_Serial-release
  PASSED clang-3.9.0-Pthread_Serial-release
Testing compiler cuda/9.1
  Starting job clang-6.0-Cuda_Pthread-release
  PASSED clang-6.0-Cuda_Pthread-release
  Starting job cuda-9.1-Cuda_OpenMP-release
  PASSED cuda-9.1-Cuda_OpenMP-release
#######################################################
PASSED TESTS
#######################################################
clang-3.9.0-Pthread_Serial-release build_time=125 run_time=428
clang-6.0-Cuda_Pthread-release build_time=211 run_time=307
cuda-9.1-Cuda_OpenMP-release build_time=265 run_time=195
gcc-4.8.4-OpenMP-release build_time=95 run_time=133
gcc-4.8.4-Pthread-release build_time=90 run_time=304
gcc-5.3.0-Serial-release build_time=104 run_time=164
intel-16.0.1-OpenMP-release build_time=307 run_time=153
```

**Comment by mhoemmen** [comment]
@micahahoward @rrdrake

**Pull request merged**